### ISSUE TITLE
AZW3 Support?

### ISSUE THREAD

**Issue by edent** [koreader/koreader#5845]
* KOReader version: 2020.01
* Device: Likebook Ares

#### Issue
Can't open AZW3 files (without DRM).

#### Steps to reproduce
Try to open a correctly formatted AZW3 file. Doesn't open. Renamed the extension to `.mobi` - still doesn't open.

I know I can manually convert using Calibre - but is there any way to get KOReader to open AZW3?

**Comment by Frenzie** [comment]
Did you try what happens when you rename it to .azw? If it's reasonable then we can add it to the list of "supported" file formats so it shows up in the file manager. Even so results are likely to be better when converted to EPUB by an application like calibre.

https://github.com/koreader/koreader/blob/e49198cef85f939f1a416291a6ea1ba0dabf5565/frontend/document/credocument.lua#L858

**Comment by edent** [comment]
Ah, I saw that [Librera for Android](https://github.com/foobnix/LibreraReader) supported AZW3, so thought it might be simple.

If I rename the file to azw it shows this mess :-)
![Screenshot_20200211-132926](https://user-images.githubusercontent.com/837136/74241186-671d3800-4cd3-11ea-9c89-03f3a702001a.png)

**Comment by Frenzie** [comment]
> Ah, I saw that Librera for Android supported AZW3, so thought it might be simple.

If it is, it'd only be because calibre/Librera/someone already reverse engineered the proprietary format. ;-) But the kind of simple I'm talking about in that case is more like only 10 hours instead of 100 hours. (Not an informed estimate by any means but I think you'll get the point.)

> If I rename the file to azw it shows this mess :-)

Is that the same throughout though, or are there parts that do show up correctly?

Depending on what they changed it's possible that a relatively simple fix along these lines could be sufficient, for example:
https://github.com/koreader/crengine/commit/28fefeb4115e9e1f24c98edeb33834a3ae681df4

**Comment by poire-z** [comment]
I don't know if AZW3 is still the kind of MOBI crengine can decode.
Anyway, the 2 samples I had working with the above fix, when opened with vim, start with the title and some string BOOKMOBI not far away:
![image](https://user-images.githubusercontent.com/24273478/74244059-23c9c600-4ce1-11ea-93a2-45fa06336191.png)
and a bit later, another visible string "MOBI" that the code expects to find at some point among the binary headers:
![image](https://user-images.githubusercontent.com/24273478/74244140-452ab200-4ce1-11ea-9183-035782bd29c8.png)

If your book kinda looks like that, there are chances it's just some header tweaks.
If not, it's probably another container format, and there's not much chances you'll get it working with crengine.

**Comment by edent** [comment]
My books have those strings near the start! `BOOKMOBI` and `MOBI`.

I'm happy to binary edit them to test if the engine can do something with them. Or if you have a build of crengine with updated header parser?

**Comment by poire-z** [comment]
Well, there's been no update on that front that you wouldn't already have in KOReader release or nightly.
You'd need to build the emulator to be able to debug the code and see where it fails... :|

All the fun must be happening in that file https://github.com/koreader/crengine/blob/master/crengine/src/pdbfmt.cpp
I haven't really understand how all that works, and the only time I try to solve an issue, I was just putting some printf here and there.
But may be by looking at the binary of your file, and trying to map what you see to the expected header bits' size and order (not fun), you'll see something.

**Comment by edent** [comment]
Aha! I think I have an answer.

* According to [MobiRead Wiki](https://wiki.mobileread.com/wiki/AZW3) the AZW3 format is an ePub wrapped in a Mobibook header.
* Using the [KindleUnpack](https://www.mobileread.com/forums/showthread.php?t=61986) python script, I extracted the ePub. I don't think this does any conversion, it's just a straight copy.
* I copied the ePub across to my eReader, and KOReader opened it perfectly!

So, it *looks* like using https://github.com/kevinhendricks/KindleUnpack it should be possible to just read the xhtml directly from the AZW3.

The header for the AZW3 file is:

```
Huffdic compression
Dumping section 0, Mobipocket Header version: 8, total length 280
Mobipocket header from section 0
     Offset  Value Hex Dec        Description
0x000 (  0)     0x4448      17480 compression_type
0x002 (  2)     0x0000          0 fill0
0x004 (  4) 0x00098224     623140 text_length
0x008 (  8)     0x009C        156 text_records
0x00A ( 10)     0x1000       4096 max_section_size
0x00C ( 12)     0x0000          0 crypto_type
0x00E ( 14)     0x0000          0 fill1
0x010 ( 16)       MOBI            magic
0x014 ( 20) 0x00000108        264 header_length (from MOBI)
0x018 ( 24) 0x00000002          2 type
0x01C ( 28) 0x0000FDE9      65001 codepage
0x020 ( 32) 0xFC8C034E 4237034318 unique_id
0x024 ( 36) 0x00000008          8 version
0x028 ( 40) 0x0000009E        158 metaorthindex
0x02C ( 44) 0xFFFFFFFF 4294967295 metainflindex
0x030 ( 48) 0xFFFFFFFF 4294967295 index_names
0x034 ( 52) 0xFFFFFFFF 4294967295 index_keys
0x038 ( 56) 0xFFFFFFFF 4294967295 extra_index0
0x03C ( 60) 0xFFFFFFFF 4294967295 extra_index1
0x040 ( 64) 0xFFFFFFFF 4294967295 extra_index2
0x044 ( 68) 0xFFFFFFFF 4294967295 extra_index3
0x048 ( 72) 0xFFFFFFFF 4294967295 extra_index4
0x04C ( 76) 0xFFFFFFFF 4294967295 extra_index5
0x050 ( 80) 0x0000009E        158 first_nontext
0x054 ( 84) 0x000005CC       1484 title_offset
0x058 ( 88) 0x00000015         21 title_length
0x05C ( 92) 0x00000009          9 language_code
0x060 ( 96) 0x00000000          0 dict_in_lang
0x064 (100) 0x00000000          0 dict_out_lang
0x068 (104) 0x00000008          8 min_version
0x06C (108) 0x000000AC        172 first_resc_offset
0x070 (112) 0x000000A9        169 huff_offset
0x074 (116) 0x00000003          3 huff_num
0x078 (120) 0x000000B3        179 huff_tbl_offset
0x07C (124) 0x00000001          1 huff_tbl_len
0x080 (128) 0x00000050         80 exth_flags
0x084 (132) 0x00000000          0 fill3_a
0x088 (136) 0x00000000          0 fill3_b
0x08C (140) 0x00000000          0 fill3_c
0x090 (144) 0x00000000          0 fill3_d
0x094 (148) 0x00000000          0 fill3_e
0x098 (152) 0x00000000          0 fill3_f
0x09C (156) 0x00000000          0 fill3_g
0x0A0 (160) 0x00000000          0 fill3_h
0x0A4 (164) 0xFFFFFFFF 4294967295 unknown0
0x0A8 (168) 0xFFFFFFFF 4294967295 drm_offset
0x0AC (172) 0x00000000          0 drm_count
0x0B0 (176) 0x00000000          0 drm_size
0x0B4 (180) 0x00000000          0 drm_flags
0x0B8 (184) 0x00000000          0 fill4_a
0x0BC (188) 0x00000000          0 fill4_b
0x0C0 (192) 0x000000B4        180 fdst_offset
0x0C4 (196) 0x00000002          2 fdst_flow_count
0x0C8 (200) 0x000000B6        182 fcis_offset
0x0CC (204) 0x00000001          1 fcis_count
0x0D0 (208) 0x000000B5        181 flis_offset
0x0D4 (212) 0x00000001          1 flis_count
0x0D8 (216) 0x00000000          0 unknown1
0x0DC (220) 0x00000000          0 unknown2
0x0E0 (224) 0xFFFFFFFF 4294967295 srcs_offset
0x0E4 (228) 0x00000000          0 srcs_count
0x0E8 (232) 0xFFFFFFFF 4294967295 unknown3
0x0EC (236) 0xFFFFFFFF 4294967295 unknown4
0x0F0 (240)     0x0000          0 fill5
0x0F2 (242)     0x0003          3 traildata_flags
0x0F4 (244) 0x000000A6        166 ncx_index
0x0F8 (248) 0x0000009E        158 fragment_index
0x0FC (252) 0x000000A1        161 skeleton_index
0x100 (256) 0x000000B7        183 datp_offset
0x104 (260) 0x000000A3        163 guide_index
```

**Comment by NiLuJe** [comment]
Short answer: No.

Long answer: AZW3 is the filename given to *all* the KF8 variants. If the files comes straight from kindlegen, it does indeed contain the ePub source. But then if you have one of those, it generally means you're the content producer, so just use the ePub source ;).

Because otherwise, it'll bundle either the KF8 ML, or the KF8 ML *AND* the legacy M7 ML. Files shipped to a KF8-capable devices are stripped down to only the KF8 ML, files shipped to non-KF8 capable devices are downgraded to plain old M7 (i.e., AZW).

TL;DR: Either use the ePub source, or pass it through KindleUnpack to either get the old M7 version out, or, better, the ePub conversion it does of the KF8 bits.

**Comment by NiLuJe** [comment]
Otherwise, my answer is the same as the last time this has been brought up:

[Don't reinvent the wheel](https://github.com/bfabiszewski/libmobi), PRs welcome.

(All the implementation I know of are either based on KU, Calibre, or libmobi. For good reason).

**Comment by poire-z** [comment]
For the record (not interested in implementing it, as I never had any AZW):

That thornyreader fork https://github.com/koreader/crengine/issues/267 was using libmobi.
But what it did was a bit ugly:
```
// detect and read the file with libmobi:
MOBIData *m = mobi_init();
MOBI_RET mobi_ret = mobi_load_file(m, file);
MOBIRawml *rawml = mobi_init_rawml(m)
mobi_ret = mobi_parse_rawml(rawml, m)
// all fine till here :) and then :(
create_epub(rawml,epubnewpath)
```
The last thingy is a custom function based on https://github.com/bfabiszewski/libmobi/blob/public/tools/mobitool.c#L504, that creates a tmp.epub, which is then opened by crengine EPUB code.

The right, fun, and pretty way to go at it would be to copy epubfmt.cpp to mobifmt.cpp, and use the rawml directly at the equivalent places to feed crengine generic document building function from the rawml bits directly, avoiding the uneeded `save to EPUB > load from EPUB` steps.

**Comment by Frenzie** [comment]
Alright, I'll close it since none of us will work on it. Our position was expressed by @NiLuJe in https://github.com/koreader/koreader/issues/5845#issuecomment-584724832 PRs welcome though. ;-)

**Comment by NiLuJe** [comment]
@poire-z: Yeah, that's often (unfortunately) the way this ends up being done. (e.g., foliate did the same via KU last time I checked).

**Comment by johnd0e** [comment]
@NiLuJe 

> But then if you have one of those, it generally means you're the content producer, so just use the ePub source ;).

As I see **all** my azw3 books came from [sendtokindle](https://www.amazon.com/gp/sendtokindle).
So technically you are right, as I could use another delivery methods instead to keep original books formats (mobi, epub).

But.. **sendtokindle** services are really convenient.
